[WIP] Initial project setup

kuries opened the pull request:
I have added the boilerplate code for the two services Frontend (React) and Backend (FastAPI) including the configuration file for the Nginx reverse proxy.

Support for docker has been added to run all the three services in development mode.

Half-Shot commented on backend/Dockerfile:
```diff
@@ -0,0 +1,13 @@
+FROM python:3.8
```
I always advise latest and greatest stable for Docker images :)

kuries replied:
Changed it to 3.10 in the commit [8546e2c](https://github.com/MatrixDoctors/matrix-cerberus/pull/1/commits/8546e2ca34db5ee5dc47a7a5221c5ec5c26b94ae)

Half-Shot commented on nginx/dev/nginx.conf:
```diff
@@ -0,0 +1,39 @@
+			proxy_set_header Connection "upgrade";
+		}
+
+		location /api {
```
You may need to set some sane defaults for CORS https://enable-cors.org/server_nginx.html

kuries replied:
Will we be needing this if we host both the services under the same domain?

Half-Shot commented on docker-compose.yml:
```diff
@@ -0,0 +1,32 @@
+    expose:
+      - 3000
+    environment:
+      - NODE_ENV=development
```
Do we want to hardcode the env to be development here? Feels unexpected if you are trying to run this project in production?

Half-Shot commented on frontend/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+RUN npm ci
+
+# eslint is present in .cache
+RUN mkdir node_modules/.cache && chmod -R 777 node_modules/.cache
```
This seems very funky. You probably don't want to be installing eslint at all?

Half-Shot commented on frontend/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+# install and cache app dependencies
+COPY package.json .
+COPY package-lock.json .
+RUN npm ci
```
```suggestion
RUN npm ci && npm cache clean
```

This cleans out the cache after an install, which reduces image bloat. We do something similar [here](https://github.com/matrix-org/matrix-hookshot/blob/main/Dockerfile#L33)

Half-Shot commented on frontend/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+# pull official base image
+FROM node:16.13.2-alpine
+
```
```suggestion
NODE_ENV=production
```

So we build production things and don't install dev dependencies.

kuries replied:
Done 👍

Half-Shot commented on frontend/README.md:
```diff
@@ -0,0 +1,70 @@
+# Getting Started with Create React App
```
We should probably delete most of this and stick up a boiler plate "Hello this is the frontend, read the ../README.md" please.

Half-Shot commented on nginx/dev/Dockerfile:
```diff
@@ -0,0 +1,2 @@
+FROM nginx:latest
+COPY ./nginx.conf /etc/nginx/nginx.conf
```
I'm wondering if it's not easier to just mount the config at runtime to save ourselves building an image?

Half-Shot commented on frontend/src/index.js:
```diff
@@ -0,0 +1,17 @@
+// If you want to start measuring performance in your app, pass a function
+// to log results (for example: reportWebVitals(console.log))
+// or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
+reportWebVitals();
```
I've never used this before, so I have no advice for it. Are we going to be using it?

kuries commented:
@Half-Shot 
I've updated my docker-compose and its respective Dockerfiles to support production. I also added the CORS defaults for Nginx and removed the Nginx development configuration for now.

Could you please go over the changes and give suggestions for the same?

Cadair reviewed (approved):
Python side looks good.

Half-Shot reviewed (approved):
Looks solid to me :+1:

Cadair commented:
:shipit:
